pkg/ui/server.go

```diff
@@ -0,0 +1,282 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ui
+
+import (
+	"context"
+	"fmt"
+	"html/template"
+	"log"
+	"net/http"
+	"strings"
+)
+
+// Pair represents a key value pair used by the select HTML element.
+type Pair struct {
+	Key  string
+	Text string
+}
+
+// Content defines the displayable parts of the token retrieval form.
+type Content struct {
+	UserLabel     string
+	CategoryLabel string
+	ReasonLabel   string
+	TTLLabel      string
+	Categories    []Pair
+	TTLs          []Pair
+}
+
+// FormDetails represents all the input and content used for the token retrievlal form.
+type FormDetails struct {
+	WindowName string
+	Origin     string
+	PageTitle  string
+	Content    Content
+	Category   string
+	Reason     string
+	TTL        string
+	Errors     map[string]string
+}
+
+// SuccessDetails represents the data used for the success page and the postMessage response to the client.
+type SuccessDetails struct {
+	PageTitle    string
+	Token        string
+	TargetOrigin string
+	WindowName   string
+}
+
+// ForbiddenDetails represents the data used for the forbidden page.
+type ForbiddenDetails struct {
+	PageTitle string
+	Message   string
+}
+
+// Server holds the parsed html templates.
+type Server struct {
+	templates map[string]*template.Template
+	allowList []string
+}
+
+var (
+	categories = []string{"explanation", "breakglass"}
+	ttls       = []string{"15", "30", "60", "120", "240"}
+)
+
+// NewServer creates a new HTTP server implementation that will handle
+// rendering the JVS form and parses the go templates.
+func NewServer(ctx context.Context, cfg *ServiceConfig, tmplLocations map[string]string) (*Server, error) {
+	templateMap := make(map[string]*template.Template)
+
+	// Parse templates
+	for key, path := range tmplLocations {
+		tmpl, err := template.ParseFiles(path)
+		if err != nil {
+			return nil, fmt.Errorf("parsing %s template: %w", path, err)
+		}
+		templateMap[key] = tmpl
+	}
+
+	return &Server{
+		templates: templateMap,
+		allowList: cfg.AllowList,
+	}, nil
+}
+
+// Routes creates a ServeMux of all of the routes that
+// this Router supports.
+func (s *Server) Routes() http.Handler {
+	mux := http.NewServeMux()
+	fs := http.FileServer(http.Dir("./assets/static"))
+	mux.Handle("/assets/static/", http.StripPrefix("/assets/static/", fs))
+	mux.Handle("/popup", s.handlePopup())
+	return mux
+}
+
+func (s *Server) handlePopup() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		handlePopupFunc(w, r, s.templates, s.allowList)
+	})
+}
+
+// handlePopupFunc the initial page load of the form as well as responds to the submission of the form and ultimately renders an HTML page.
+func handlePopupFunc(w http.ResponseWriter, r *http.Request, templates map[string]*template.Template, allowList []string) {
```

Comment:
Usually you would put the logic directly in the handler. This also avoids the need to parameterize things passed to another function:

```go
func (s *Server) HandlePopup() http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    formDetails := getFormDetails(r)

    if r.Method === http.MethodGet {

    // ..
  })
}
```